ta_services/keymaster/km.c

```diff
@@ -1,10 +1,47 @@
 /* SPDX-License-Identifier: BSD-2-Clause */
 /* Copyright (c) 2018, Linaro Limited */
 
-#include <tee_internal_api.h>
-
 #include "km.h"
 
+static bool version_info_set = false;
+static uint32_t boot_os_version = 0;
+static uint32_t boot_os_patchlevel = 0;
+
+bool km_is_configured(uint32_t cmd)
+{
+	if (cmd != KEYMASTER_CMD_CONFIGURE && !version_info_set) {
+		EMSG("Keymaster TA not configured!");
+		return false;
+	}
+	else
+		return true;
+}
+
+TEE_Result km_configure(uint32_t os_version, uint32_t os_patchlevel)
+{
+	IMSG("setting version info");
+	IMSG("os_version = %u", os_version);
+	IMSG("os_patchlevel = %u", os_patchlevel);
+
+	if (!version_info_set) {
+		/*
+		 * https://android.googlesource.com/trusty/app/keymaster/+/994293cc45700fa58512b312c94da0f46d95403e
+		 * Note that version info is now set by Configure, rather than by the bootloader.  This is
+		 * to ensure that system-only updates can be done, to avoid breaking Project Treble.
+		 */
+		boot_os_version = os_version;
+		boot_os_patchlevel = os_patchlevel;
+		version_info_set = true;
+    }
+	else {
+		IMSG("version info already set");
```

Comment:
https://source.android.com/security/keystore/implementer-ref#configure says:
> If the keymaster implementation is in secure hardware and the OS version and patch level values provided do not match the values provided to the secure hardware by the bootloader (or if the bootloader did not provide values), then this method returns KM_ERROR_INVALID_ARGUMENT, and all other methods continue returning KM_ERROR_KEYMASTER_NOT_CONFIGURED.

To me that seems like we're supposed to return an error if the value doesn't match previously supplied values.